jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java

```diff
@@ -591,10 +619,34 @@ public boolean isShutdown()
     @Override
     public CompletableFuture<Void> shutdown()
     {
-        // TODO
-        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
-        completableFuture.complete(null);
-        return completableFuture;
+        while (true)
+        {
+            Availability availability = _availability.get();
+            switch (availability)
+            {
+                case STOPPED ->
+                {
+                    return CompletableFuture.completedFuture(null);
+                }
+                case SHUTDOWN ->
+                {
+                    return _shutdown;
+                }
+                default ->
+                {
+                    if (!_availability.compareAndSet(availability, Availability.SHUTDOWN))
+                        continue;
+                    checkShutdown();
+                    return _shutdown;
+                }
+            }
+        }
+    }
+
+    protected void checkShutdown()
+    {
+        if (_graceful && !_shutdown.isDone() && _requests.sum() == 0)
```

Comment:
Isn't this mechanism of counting the requests a duplicate of what `GracefulHandler` already does?
The idea was to extract it out from `StatisticsHandler` and have just `GracefulHandler` to all the work, no?
Why do we need it also here?